lib/widgets/swipable_message_row.dart

```diff
@@ -0,0 +1,80 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import '../api/model/model.dart';
+import 'icons.dart';
+import 'text.dart';
+import 'theme.dart';
+
+class SwipableMessageRow extends StatefulWidget {
+  const SwipableMessageRow({
+    super.key,
+    required this.children,
+    required this.message,
+  });
+
+  final List<Widget> children;
+  final Message message;
+
+  @override
+  State<StatefulWidget> createState() => _SwipableMessageRowState();
+}
+
+class _SwipableMessageRowState extends State<SwipableMessageRow> {
+  @override
+  Widget build(BuildContext context) {
+    final hasMarker = widget.message.editState != MessageEditState.none;
+
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.baseline,
+      textBaseline: localizedTextBaseline(context),
+      children: [
+        hasMarker
+          ? _EditStateMarker(editState: widget.message.editState)
+          : const SizedBox(width: 16),
+        ...widget.children,
+      ],
+    );
+  }
+}
+
+class _EditStateMarker extends StatelessWidget {
+  /// The minimum width of the marker.
+  // Currently, only the collapsed state of the marker has been implemented,
+  // where only the marker icon, not the marker text, is visible.
+  static const double widthCollapsed = 16;
+
+  const _EditStateMarker({
+    required MessageEditState editState,
+  }) : _editState = editState;
+
+  final MessageEditState _editState;
+
+  @override
+  Widget build(BuildContext context) {
+    final designVariables = DesignVariables.of(context);
+
+    final IconData icon;
+    final double iconSize;
+
+    switch (_editState) {
```

Comment:
nit: these two locals are tightly related to this `switch` (it's what sets them), so keep them adjacent in the same stanza:
```suggestion
    final IconData icon;
    final double iconSize;
    switch (_editState) {
```